plugins/aggregators/derivative/derivative.go

```diff
@@ -0,0 +1,225 @@
+	# suffix = "_by_time"
+	##
+	## As an abstraction the derivative can be calculated not only by the time
+	## difference but by the difference of a field, which is contained in the
+	## measurement. This field is assumed to be monotonously increasing. This
+	## feature is used by specifying a *variable*.
+	## Make sure the specified variable is not filtered and exists in the metrics 
+	## passed to this aggregator!
+	# variable = ""
+	##
+	## When using a field as the derivation parameter the name of that field will
+	## be used for the resulting derivative, e.g. *fieldname_by_parameter*.
+	##
+	## Note, that the calculation is based on the actual timestamp of the
+	## measurements. When there is only one measurement during that period, the
+	## measurement will be rolled over to the next period. The maximum number of
+	## such roll-overs can be configured with a default of 10.
+	# max_roll_over = 10
+	##
+`
+
+func (d *Derivative) SampleConfig() string {
+	return sampleConfig
+}
+
+func (d *Derivative) Description() string {
+	return "Calculates a derivative for every field."
+}
+
+func (d *Derivative) Add(in telegraf.Metric) {
+	id := in.HashID()
+	if current, ok := d.cache[id]; !ok {
+		// hit an uncached metric, create caches for first time:
+		d.cache[id] = newAggregate(in)
+	} else {
+		if current.first.time.After(in.Time()) {
+			current.first = newEvent(in)
+			current.rollOver = 0
+		} else if current.first.time.Equal(in.Time()) {
+			upsertConvertedFields(in.Fields(), current.first.fields)
+			current.rollOver = 0
+		}
+		if current.last.time.Before(in.Time()) {
+			current.last = newEvent(in)
+			current.rollOver = 0
+		} else if current.last.time.Equal(in.Time()) {
+			upsertConvertedFields(in.Fields(), current.last.fields)
+			current.rollOver = 0
+		}
+	}
+}
+
+func newAggregate(in telegraf.Metric) *aggregate {
+	event := newEvent(in)
+	return &aggregate{
+		name:     in.Name(),
+		tags:     in.Tags(),
+		first:    event,
+		last:     event,
+		rollOver: 0,
+	}
+}
+
+func newEvent(in telegraf.Metric) *event {
+	return &event{
+		fields: extractConvertedFields(in),
+		time:   in.Time(),
+	}
+}
+
+func extractConvertedFields(in telegraf.Metric) map[string]float64 {
+	fields := make(map[string]float64, len(in.Fields()))
+	upsertConvertedFields(in.Fields(), fields)
+	return fields
+}
+
+func upsertConvertedFields(source map[string]interface{}, target map[string]float64) {
+	for k, v := range source {
+		if value, ok := convert(v); ok {
+			target[k] = value
+		}
+	}
+}
+
+func convert(in interface{}) (float64, bool) {
+	switch v := in.(type) {
+	case float64:
+		return v, true
+	case int64:
+		return float64(v), true
+	case uint64:
+		return float64(v), true
+	}
+	return 0, false
+}
+
+func (d *Derivative) Push(acc telegraf.Accumulator) {
+	for _, aggregate := range d.cache {
+		if aggregate.first == aggregate.last {
+			d.Log.Warnf("Same first and last event for %q, skipping.", aggregate.name)
+			continue
+		}
```

Comment:
wouldn't this be a 0 value derivative result?

Reply:
It is undefined, since the denominator would be 0.

Reply:
Furthermore, notice, that the value can be "rolled-over" to the next interval, where a proper value may be calculated without emitting a 0 in between. That is actually one of the goals of the roll-over.